mmdet3d/models/necks/view_transformer.py

```diff
@@ -0,0 +1,310 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+from mmcv.runner import BaseModule
+
+from ..builder import NECKS
+
+
+class QuickCumsum(torch.autograd.Function):
+    """Back propagation accelerated cumulative sum operator."""
+
+    @staticmethod
+    def forward(ctx, x, coor, ranks):
+        x = x.cumsum(0)
+        kept = torch.ones(x.shape[0], device=x.device, dtype=torch.bool)
+        kept[:-1] = (ranks[1:] != ranks[:-1])
+
+        x, coor = x[kept], coor[kept]
+        x = torch.cat((x[:1], x[1:] - x[:-1]))
+
+        # save kept for backward
+        ctx.save_for_backward(kept)
+
+        # no gradient for coordinates
+        ctx.mark_non_differentiable(coor)
+
+        return x, coor
+
+    @staticmethod
+    def backward(ctx, gradx, gradgeom):
+        kept, = ctx.saved_tensors
+        back = torch.cumsum(kept, 0)
+        back[kept] -= 1
+
+        val = gradx[back]
+
+        return val, None, None
+
+
+@NECKS.register_module()
+class ViewTransformerLiftSplatShoot(BaseModule):
+    """Lift-Splat-Shoot view transformer for transform image-view feature into
+    bird-eye-view feature.
+
+    Args:
+        grid_config (dict(axis:list(float,float,float))): Config of grid alone
+            each axis in format of (lower_bound, upper_bound, interval). axis
+            in {x,y,z,depth}.
+        input_size (tuple(int)): Size of input images in format of (height,
+            width).
+        in_channels (int): Channels of input feature.
+        tran_channels (int): Channels of transformed feature.
+        accelerate (bool): Whether the view transformation is conducted with
+            acceleration. Note: the intrinsic and extrinsic of cameras should
+            be constant when 'accelerate' is set true.
+        max_voxel_points (int): Specify the maximum point number in a single
+            voxel during the acceleration.
+    """
+
+    def __init__(self,
+                 grid_config,
+                 input_size,
+                 downsample,
+                 in_channels,
+                 tran_channels,
+                 accelerate=False,
+                 max_voxel_points=300):
+        super(ViewTransformerLiftSplatShoot, self).__init__()
+        self.grid_config = grid_config
+        self.gen_grid_infos(**self.grid_config)
+
+        self.input_size = input_size
+        self.downsample = downsample
+
+        self.create_frustum()
+        self.tran_channels = tran_channels
+        self.depthnet = nn.Conv2d(
+            in_channels, self.D + self.tran_channels, kernel_size=1, padding=0)
+        self.accelerate = accelerate
+        self.max_voxel_points = max_voxel_points
+        self.initial_flag = True
+
+    def gen_grid_infos(self, x, y, z, **kwargs):
```

Comment:
Why use such a method to init the grid info

Reply:
Is there any big difference from initializing the grid with simple xxx.new_tensor/torch.tensor?